lib/schedule/block.ex

```diff
@@ -1,101 +1,95 @@
 defmodule Schedule.Block do
+  alias Schedule.AsDirected
   alias Schedule.Gtfs.Service
   alias Schedule.Hastus
+  alias Schedule.Piece
   alias Schedule.Trip
 
+  require Logger
+
   @type id :: String.t()
   @typedoc """
   Block ids are repeated between different services.
-  In order to uniquely identify a block, you need a Service.id() in addition to a Block.id()
+  In order to uniquely identify a block, you need a Schedule.id() in addition to a Block.id()
   """
-  @type key :: {id(), Service.id()}
+  @type key :: {Hastus.Schedule.id(), id()}
   @type by_id :: %{key() => t()}
 
   @type t :: %__MODULE__{
           id: id(),
           service_id: Service.id(),
```

Comment:
Is `service_id` still guaranteed to be non-nil? In practice I'm guessing it always is non-nil, but you do theoretically have the case of a block that's all nonrevenue trips that don't exist in GTFS.

Reply:
So I can't 100% say "guaranteed", but I can say I did look through the data, and saw no such blocks. If you think about it, that would correspond to pulling the bus out, driving around for a while without picking up a single passenger, then pulling back to the garage, and I don't see why we would ever do that. Maybe @dks-mbta could say definitively if there's a valid reason to have a block like that.

Generally, apart from non-revenue trips, there is one other important case to be aware of where a trip might have a nil `service_id` (maybe you're aware already). That's when the trip is in the data but is for a rating other than the current one, and we do filter that out via `Piece.from_non_current_rating?/1`, which also has a big comment by it explaining why those trips are there.

Reply:
Yeah, I guess this gets at a question of whether our types should reflect the desired shape of the data or, like, all possible values of the data. Perhaps in a more strongly-typed language we would solve this by having a function from a "all possible states of the input" type to a "desired internal data representation OR error" type that does the relevant parsing, but I suppose we're not writing in such a language and, as you point out, Gleam is not yet production ready. I guess this is fine as-is.